### PR TITLE
Feature/hyperledger chaincode courses

### PR DESCRIPTION
**Pull request by NikoBergemann**
### Reason for this PR
- We want to deploy the hyperledger network to the IRB server to set up our MVP

### Changes in this PR
- Provide Hyperledger Network to work on any (sufficient) HostSystem
- Provide Hyperledger Chaincode to store course-data
- Provide Hyperledger Scala API to access ChainCode
- Provide HyperledgerService to access Chaincode per Lagom ServiceAPI
- Provide HyperledgerCourseService to enable CourseManagement via a dedicated LagomService
- Provide Unit-/Module-/ and Integration-Tests to ensure Hyperledger Networks behaves as specified

### CODE REVIEW

**Comment by NikoBergemann** [comment]
@ Lagom-people: Der HyperledgerCourseService hat bisher addCourse(Course, Done) implementiert. Nach merge from Develop ist das offizielle CourseInterface allerdings addCourse(course, course). Was ist hier die richtige herangehensweise?

File: product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  /** @inheritdoc */
+  override def addCourse(): ServiceCall[Course, Done] =
+    identifiedAuthenticated(AuthenticationRole.Admin, AuthenticationRole.Lecturer) {
+      (username, role) =>
+        ServerServiceCall {
+          (_,courseProposal) =>
+            if (role == AuthenticationRole.Lecturer && courseProposal.lecturerId.trim != username){
+              throw new CustomException(TransportErrorCode(403, 1003, "Error"), DetailedError("owner mismatch", List()))
+            }
+            val courseToAdd = courseProposal.copy(courseId =java.util.UUID.randomUUID().toString)
```
Comment (david-buderus) [inline product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala:43]: Change UUID Generation to time-based like in CourseServiceImpl

File: product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            val validationErrors = validateCourseSyntax(courseToAdd)
+            if (validationErrors.nonEmpty) {
+              throw new CustomException(TransportErrorCode(422, 1003, "Error"), DetailedError("validation error", validationErrors))
+            }
+            hyperLedgerSession.write("addCourse", courseToAdd).map((ResponseHeader(201, MessageProtocol.empty, List()), _))
```
Comment (david-buderus) [inline product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala:48]: Return courseToAdd after(!) the write finishes

File: product_code/hyperledger/api/src/test/scala/ChaincodeConnectionObjectTests.scala [hunk 3]
```diff
@@ -1,66 +1,89 @@
...
-    Using(ConnectionManager.createConnection()) { chaincodeConnection: ChaincodeTrait =>
-      // initial courses
-      val courses = chaincodeConnection.getAllCourses()
-      assert(courses != null, "Get All courses returned null")
-      println("Courses: " + courses)
+      "Allow a full walkthrough" in {
+        val testResult = Using(connectionManager.createConnection()) { chaincodeConnection: ChaincodeTrait =>
+          // initial courses
+          val courses = chaincodeConnection.getAllCourses()
+          assert(courses != null, "Get All courses returned null")
```
Comment (david-buderus) [inline product_code/hyperledger/api/src/test/scala/ChaincodeConnectionObjectTests.scala:35]: We agreed to use Matchers with WordSpec for Testing

File: product_code/hyperledger/api/src/test/scala/ChaincodeQuickAccessTests.scala [hunk 4]
```diff
@@ -1,17 +1,20 @@
...
 
-    // debug help
-    println(result)
+        // debug help
+        println(result)
 
-    // perform tests
-    assert(result != null, "result was null")
+        // perform tests
+        assert(result != null, "result was null")
```
Comment (david-buderus) [inline product_code/hyperledger/api/src/test/scala/ChaincodeQuickAccessTests.scala:16]: We agreed to use Matchers with WordSpec for Testing

File: product_code/hyperledger/api/src/test/scala/ConnectionTests.scala [hunk 5]
```diff
@@ -1,64 +1,62 @@
...
+    "connecting to Chain" should {
+
+      "provide gateway connection" in {
+        // retrieve possible identities
+        val wallet = connectionManager.getWallet()
+        assert(wallet != null, "Wallet retrieved was null.")
+
+        // prepare Network Builder
+        val builder = connectionManager.getBuilder(wallet)
+        assert(builder != null, "Builder retrieved was null, maybe the connection profile did not match the running network.")
```
Comment (david-buderus) [inline product_code/hyperledger/api/src/test/scala/ConnectionTests.scala:19]: We agreed to use Matchers with WordSpec for Testing

File: product_code/lagom/build.sbt [hunk 6]
```diff
@@ -45,14 +45,32 @@ val defaultPersistenceKafkaDependencies = Seq(
   lagomScaladslKafkaBroker,
 )
 
+/*
```
Comment (david-buderus) [inline product_code/lagom/build.sbt:48]: Remove unused code

File: product_code/lagom/hl_course_service/impl/src/main/resources/application.conf [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#
+#
+play.application.loader = de.upb.cs.uc4.hl.course.impl.HlCourseLoader
+
+db.default {
```
Comment (david-buderus) [inline product_code/lagom/hl_course_service/impl/src/main/resources/application.conf:5]: The hl_course_service doesn't use Postgres, no driver definition needed

File: product_code/lagom/hl_course_service/impl/src/main/resources/prod-application.conf [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  http.secret.key = "${APPLICATION_SECRET}"
+}
+
+akka {
+  discovery.method = akka-dns
+
+  management.cluster.bootstrap {
+    contact-point-discovery {
+      discovery-method = kubernetes-api
+      service-name = "hl_course"
```
Comment (david-buderus) [inline product_code/lagom/hl_course_service/impl/src/main/resources/prod-application.conf:17]: The service is called hlcourse  
Change it here or in the HlCourseService.scala

File: product_code/lagom/hl_course_service/impl/src/main/resources/application.conf (hunk 7 above)
Comment (NikoBergemann) [reply product_code/lagom/hl_course_service/impl/src/main/resources/application.conf:5]: Remove whole db.default?

Comment (david-buderus) [reply product_code/lagom/hl_course_service/impl/src/main/resources/application.conf:5]: Yes

Comment (NikoBergemann) [reply product_code/lagom/hl_course_service/impl/src/main/resources/application.conf]: removed by new commit

File: product_code/hyperledger/api/src/test/scala/ConnectionTests.scala (hunk 5 above)
Comment (NikoBergemann) [reply product_code/hyperledger/api/src/test/scala/ConnectionTests.scala]: fixed in [https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896](https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896)

File: product_code/hyperledger/api/src/test/scala/ChaincodeQuickAccessTests.scala (hunk 4 above)
Comment (NikoBergemann) [reply product_code/hyperledger/api/src/test/scala/ChaincodeQuickAccessTests.scala]: fixed in [https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896](https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896)

File: product_code/hyperledger/api/src/test/scala/ChaincodeConnectionObjectTests.scala (hunk 3 above)
Comment (NikoBergemann) [reply product_code/hyperledger/api/src/test/scala/ChaincodeConnectionObjectTests.scala]: fixed in [https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896](https://github.com/upb-uc4/University-Credits-4.0/pull/121/commits/710923f1f36afa50fc8faff88e18af10d9ceb896)

File: product_code/lagom/hl_course_service/impl/src/main/resources/application.conf [hunk 9]
```diff
@@ -0,0 +1,34 @@
+#
+#
+play.application.loader = de.upb.cs.uc4.hl.course.impl.HlCourseLoader
+
+jdbc-defaults.slick.profile = "slick.jdbc.PostgresProfile$"
```
Comment (david-buderus) [inline product_code/lagom/hl_course_service/impl/src/main/resources/application.conf:5]: HlCourseService does not use a database, the database profile is not needed

Comment (NikoBergemann) [reply product_code/lagom/hl_course_service/impl/src/main/resources/application.conf]: solved in new commit

File: product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+  /** @inheritdoc */
+  override def allowedMethodsGETPUTDELETE: ServiceCall[NotUsed, Done] = allowedMethodsCustom("GET, PUT, DELETE")
+
+  /** Checks if the course attributes correspond to agreed syntax and semantics
+    *
+    * @param course which attributes shall be verified
+    * @return response-code which gives detailed description of syntax or semantics violation
+    */
+  def validateCourseSyntax(course: Course): Seq[SimpleError] = {
```
Comment (dhelmert) [inline product_code/lagom/hl_course_service/impl/src/main/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceImpl.scala:125]: This method is completely copied from CourseState, maybe put it in one place

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      }
+
+      override lazy val hyperLedgerService: HyperLedgerService = new HyperLedgerService {
+        override def write(transactionId: String): ServiceCall[Seq[String], Done] = ServiceCall { seq =>
+          transactionId match {
+            case "deleteCourseById" =>
+              if (Seq("0", "1").contains(seq.head)) {
+                Future.successful(Done)
+              } else {
+                throw BadRequest("ERROR")
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:47]: outdated... new Responsetypes maybe?

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+              }
+            case "updateCourseById"  =>
+              if (Seq("0", "1").contains(seq.head)) {
+                Future.successful(Done)
+              } else {
+                throw BadRequest("ERROR")
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:53]: outdated... new responsetypes maybe?

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        override def read(transactionId: String): ServiceCall[Seq[String], String] = ServiceCall { seq =>
+          transactionId match {
+            case "getAllCourses" => Future.successful("[]")
+            case "getCourseById" => seq.head match {
+              case course0.courseId => Future.successful(Json.stringify(Json.toJson(course0)))
+              case course2.courseId => Future.successful(Json.stringify(Json.toJson(course2)))
+              case course3.courseId => Future.successful(Json.stringify(Json.toJson(course0)))
+              case _ => throw BadRequest("ERROR")
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:66]: outdated... new responsetypes maybe?

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        answer.courseLanguage should ===(course2.courseLanguage)
+        answer.courseType should ===(course2.courseType)
+        answer.currentParticipants should ===(course2.currentParticipants)
+      }
+    }
+
+    "delete a non-existing course" in {
+      client.deleteCourse("42").handleRequestHeader(addAuthenticationHeader()).invoke().failed.map {
+        answer =>
+          answer.asInstanceOf[TransportException].errorCode.http should ===(400)
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:131]: Use custom exceptions and the respective error codes.

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+      }
+    }
+
+    "find a non-existing course" in {
+      client.findCourseByCourseId("42").handleRequestHeader(addAuthenticationHeader()).invoke().failed.map { answer =>
+        answer.asInstanceOf[TransportException].errorCode.http should ===(400)
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:137]: Use custom exceptions and the respective error codes.

File: product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+      }
+    }
+
+    "update a non-existing course" in {
+      client.updateCourse(course3.courseId).handleRequestHeader(addAuthenticationHeader()).invoke(course3).failed.map {
+        answer =>
+          answer.asInstanceOf[TransportException].errorCode.http should ===(400)
```
Comment (RHelmert) [inline product_code/lagom/hl_course_service/impl/src/test/scala/de/upb/cs/uc4/hl/course/impl/HlCourseServiceSpec.scala:144]: Use custom exceptions and the respective error codes.

**Pull request merged**